Registry-Dependent Potential for Interfaces of Water with Graphene

oywg11 opened the pull request:
**Summary**

Adds anisotropic interfacial potential (AIP), which can accurately describe the van der Waals interaction of the water–graphene interface. Both the standard version and the optimized version are included.

**Related Issue(s)**

<!--If this addresses an open GitHub issue for this project, please mention the issue number here, and describe the relation. Use the phrases `fixes #221` or `closes #135`, when you want an issue to be automatically closed when the pull request is merged-->

**Author(s)**

Standard Version:
Wengen Ouyang (Wuhan University, China)
Email: [w.g.ouyang@gmail.com](mailto:w.g.ouyang@gmail.com) 

Optimized Version:
Xiaohui Duan (Shandong University, China)
E-mail: [sunrise_duan@126.com](mailto:sunrise_duan@126.com) 

Ping Gao (National Supercomputing Center in Wuxi, China)
E-mail: [qdgaoping@gmail.com](mailto:qdgaoping@gmail.com)


**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

<!--Please state whether any changes in the pull request will break backward compatibility for inputs, and - if yes - explain what has been changed and why-->

**Implementation Notes**

And this pair style updates to the INTERLAYER package;
And the opt pair style updates to the OPT package;
And this pair style as a derived class of pair style of ilp/tmd to avoid code duplication


**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

This registry-dependent potential is designed for the accurate empirical description the anisotropic van der Waals interaction in water/graphene interfaces, which cannot be captured by the commonly adopted Lennard-Jones potential. Further details and benchmark tests can be found the following reference: https://pubs.acs.org/doi/10.1021/acs.jpcc.2c08464.

Commit:
add standard version of ilp_water_2dm

oywg11 committed:
add optmized verion of ilp_water_2dm

oywg11 committed:
add potential file

oywg11 committed:
add doc file

oywg11 committed:
add examples

oywg11 committed:
update doc file

oywg11 committed:
Update pair_ilp_water_2dm.rst

oywg11 committed:
add potential file that parameterized against with DMC reference data

oywg11 committed:
delete rst

oywg11 committed:
update doc file

oywg11 committed:
clean the duplicate info

oywg11 commented:
@akohlmey Dear Axel, I hope this message finds you well. Please accept my apologies for the interruption. I wanted to kindly bring to your attention that it appears there might have been an oversight or some other reason for the missed pull request. 
I understand that you have numerous tasks and responsibilities, and it's possible that this particular request may have slipped through the cracks. However, I would greatly appreciate your assistance in reviewing and addressing the pending pull request at your earliest convenience.

akohlmey commented:
Nothing gets lost here. The order and timeliness in which pull requests are processed depends on many factors. Pull requests that need a careful look and manual testing (like this one) can sit idle for extended times, if nobody has the time available to do it properly. Please note that none of the LAMMPS developers is able to work on LAMMPS full time, but rather we often have to "make time" from other projects or work on this in our free time. I for myself have found that the most effective way of using my time is to prioritize processing pull requests by how much effort it takes from my side to process it. The more effort, the longer it can sit idle. Also, bugfixes have a higher priority than adding new features. In this particular case, I have been mostly unavailable over the last 3 weeks, so (non-trivial) pull requests have accumulated.

The first issue with your submission is that it doesn't pass all the automatic tests. That usually prevents people to have a closer look, since pull requests *must* pass the automatic tests.

You pull request currently fails the coding style checks that can be run with: 
- make check-whitespace
- make check-permissions
- make check-homepage
- make check-errordocs
Any failures of those indicate violations of coding standards or conventions for LAMMPS and must be addressed.

I took a very cursory look and noticed another issue: it appears that you have copied modified files based on a significantly older version of LAMMPS over those from the current version. In the process you undo some of the maintenance changes that have happened since. That is a bad practice and very problematic since it re-introduces issues (however minor) that have been addressed already. The proper way of introducing your changes would be to extract the differences relative to the original files you have started programming from and then apply those to the current versions of the corresponding files. When using git for managing the source code (and not just to submit to GitHub), this is avoided by either regularly merging in changes from the upstream development branch or using the rebase functionality of git right before submitting the pull request. The former is easier to do, the latter can be a bit trickier but will produce a cleaner commit history.

As for newly added files that do not comply with the coding standard checks, you can in most cases do a "make fix-xxx" instead of "make check-xxx" to get an automated edit that is in 99% of the cases correct and makes the files compliant again. But those scripts are rather simple-minded and manual edits/checks may be needed. They won't restore other maintenance changes and fixes, though.

Since reviewing/fixing these kinds of issues takes additional time (and can be annoying at times since it is avoidable work), it usually causes delays of processing a pull request. I apologize if that is inconvenient to you.

oywg11 commented:
@akohlmey Many thanks for the detailed clarification, which makes me more clear on the working ways of LAMMPS developers. Thanks for pointing out the issues that existed in this pull request and providing the corresponding suggestions. I will try to fix these issues.

oywg11 committed:
revise the code files of standard version based on latest version of LAMMPS

oywg11 committed:
fix checkstyle

oywg11 committed:
fix a bug in the codes

oywg11 commented:
@akohlmey I fixed the issues of the standard version according to your suggestions. Thanks again!
@run-towards-the-future @dxhisboy will update the optimized version soon.
Please let us know if found new issues or have new suggestions.

oywg11 committed:
rename the pair style

oywg11 committed:
rename the opt pair style

oywg11 committed:
update the doc file

oywg11 committed:
fix checkstyle issues

oywg11 committed:
rename potential files

oywg11 committed:
update input files of examples

oywg11 committed:
rename log files of examples

oywg11 committed:
update the doc file

oywg11 committed:
correct a typo in the doc file

oywg11 committed:
update log files in the examples

oywg11 committed:
update the citation information

akohlmey commented on src/INTERLAYER/pair_ilp_tmd.cpp:
```diff
@@ -479,9 +481,63 @@ void PairILPTMD::calc_normal()
+            vect[0][1] = x[jH2][1] - ytp;
+            vect[0][2] = x[jH2][2] - ztp;
+          } else {
+            fprintf(screen, "jH1 jH2 = %d %d\n", jH1,jH2);
```
This is effectively debug output thus please remove the fprintf() calls and only leave the error->one() call.

oywg11 replied:
Removed

akohlmey replied:
Thanks.

akohlmey commented on src/INTERLAYER/pair_ilp_tmd.cpp:
```diff
@@ -589,12 +644,107 @@ void PairILPTMD::calc_normal()
+
+            cont = 2;
+          } else {
+            fprintf(screen, "jH1 jH2 = %d %d\n", jH1,jH2);
```
Please remove debug output, same as before...

oywg11 committed:
remove the fprintf debug outputs

oywg11 committed:
update affiliations

akohlmey commented:
> @run-towards-the-future @dxhisboy will update the optimized version soon.

Are these changes included? Is there anything still planned or pending from your side?
Or is the pull request now ready for a final review?
Please let us know.

akohlmey commented:
@oywg11 There seems to be a mismatch with function names or prototypes. When you declare a function "virtual", you usually do not overload it at the same time, since then the behavior can be very unexpected, as the polymorphic behavior does not apply to the overloaded version.  Can you please have a look?
When compiling the current code, I get this warning:
```
In file included from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:23,
                 from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_tmd_opt.h:23,
                 from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_tmd_opt.cpp:33:
/home/akohlmey/compile/lammps/src/INTERLAYER/pair_ilp_graphene_hbn.h:81:16: warning: ‘virtual void LAMMPS_NS::PairILPGrapheneHBN::calc_normal()’ was hidden [-Woverloaded-virtual=]
   81 |   virtual void calc_normal();
      |                ^~~~~~~~~~~
/home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:38:8: note:   by ‘LAMMPS_NS::PairILPGrapheneHBNOpt::calc_normal(int, int, int*, int, double*, double (*)[3], double (*)[3][3])’
   38 |   void calc_normal(int i, int itype, int *ILP_neigh, int nneigh, double *normal, double (*dnormdri)[3],
      |        ^~~~~~~~~~~
In file included from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:23,
                 from /home/akohlmey/compile/lammps/src/OPT/pair_saip_metal_opt.h:23,
                 from /home/akohlmey/compile/lammps/src/OPT/pair_saip_metal_opt.cpp:34:
/home/akohlmey/compile/lammps/src/INTERLAYER/pair_ilp_graphene_hbn.h:81:16: warning: ‘virtual void LAMMPS_NS::PairILPGrapheneHBN::calc_normal()’ was hidden [-Woverloaded-virtual=]
   81 |   virtual void calc_normal();
      |                ^~~~~~~~~~~
/home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:38:8: note:   by ‘LAMMPS_NS::PairILPGrapheneHBNOpt::calc_normal(int, int, int*, int, double*, double (*)[3], double (*)[3][3])’
   38 |   void calc_normal(int i, int itype, int *ILP_neigh, int nneigh, double *normal, double (*dnormdri)[3],
      |        ^~~~~~~~~~~
In file included from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:23,
                 from /home/akohlmey/compile/lammps/src/OPT/pair_aip_water_2dm_opt.h:23,
                 from /home/akohlmey/compile/lammps/src/OPT/pair_aip_water_2dm_opt.cpp:33:
/home/akohlmey/compile/lammps/src/INTERLAYER/pair_ilp_graphene_hbn.h:81:16: warning: ‘virtual void LAMMPS_NS::PairILPGrapheneHBN::calc_normal()’ was hidden [-Woverloaded-virtual=]
   81 |   virtual void calc_normal();
      |                ^~~~~~~~~~~
/home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:38:8: note:   by ‘LAMMPS_NS::PairILPGrapheneHBNOpt::calc_normal(int, int, int*, int, double*, double (*)[3], double (*)[3][3])’
   38 |   void calc_normal(int i, int itype, int *ILP_neigh, int nneigh, double *normal, double (*dnormdri)[3],
      |        ^~~~~~~~~~~
In file included from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:23,
                 from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.cpp:22:
/home/akohlmey/compile/lammps/src/INTERLAYER/pair_ilp_graphene_hbn.h:81:16: warning: ‘virtual void LAMMPS_NS::PairILPGrapheneHBN::calc_normal()’ was hidden [-Woverloaded-virtual=]
   81 |   virtual void calc_normal();
      |                ^~~~~~~~~~~
/home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:38:8: note:   by ‘LAMMPS_NS::PairILPGrapheneHBNOpt::calc_normal(int, int, int*, int, double*, double (*)[3], double (*)[3][3])’
   38 |   void calc_normal(int i, int itype, int *ILP_neigh, int nneigh, double *normal, double (*dnormdri)[3],
      |        ^~~~~~~~~~~
```

oywg11 committed:
rename function names in opt version

oywg11 committed:
fix a typo

oywg11 committed:
fix small format issues

oywg11 commented:
> @oywg11 There seems to be a mismatch with function names or prototypes. When you declare a function "virtual", you usually do not overload it at the same time, since then the behavior can be very unexpected, as the polymorphic behavior does not apply to the overloaded version. Can you please have a look? When compiling the current code, I get this warning:
> 
> ```
> In file included from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:23,
>                  from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_tmd_opt.h:23,
>                  from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_tmd_opt.cpp:33:
> /home/akohlmey/compile/lammps/src/INTERLAYER/pair_ilp_graphene_hbn.h:81:16: warning: ‘virtual void LAMMPS_NS::PairILPGrapheneHBN::calc_normal()’ was hidden [-Woverloaded-virtual=]
>    81 |   virtual void calc_normal();
>       |                ^~~~~~~~~~~
> /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:38:8: note:   by ‘LAMMPS_NS::PairILPGrapheneHBNOpt::calc_normal(int, int, int*, int, double*, double (*)[3], double (*)[3][3])’
>    38 |   void calc_normal(int i, int itype, int *ILP_neigh, int nneigh, double *normal, double (*dnormdri)[3],
>       |        ^~~~~~~~~~~
> In file included from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:23,
>                  from /home/akohlmey/compile/lammps/src/OPT/pair_saip_metal_opt.h:23,
>                  from /home/akohlmey/compile/lammps/src/OPT/pair_saip_metal_opt.cpp:34:
> /home/akohlmey/compile/lammps/src/INTERLAYER/pair_ilp_graphene_hbn.h:81:16: warning: ‘virtual void LAMMPS_NS::PairILPGrapheneHBN::calc_normal()’ was hidden [-Woverloaded-virtual=]
>    81 |   virtual void calc_normal();
>       |                ^~~~~~~~~~~
> /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:38:8: note:   by ‘LAMMPS_NS::PairILPGrapheneHBNOpt::calc_normal(int, int, int*, int, double*, double (*)[3], double (*)[3][3])’
>    38 |   void calc_normal(int i, int itype, int *ILP_neigh, int nneigh, double *normal, double (*dnormdri)[3],
>       |        ^~~~~~~~~~~
> In file included from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:23,
>                  from /home/akohlmey/compile/lammps/src/OPT/pair_aip_water_2dm_opt.h:23,
>                  from /home/akohlmey/compile/lammps/src/OPT/pair_aip_water_2dm_opt.cpp:33:
> /home/akohlmey/compile/lammps/src/INTERLAYER/pair_ilp_graphene_hbn.h:81:16: warning: ‘virtual void LAMMPS_NS::PairILPGrapheneHBN::calc_normal()’ was hidden [-Woverloaded-virtual=]
>    81 |   virtual void calc_normal();
>       |                ^~~~~~~~~~~
> /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:38:8: note:   by ‘LAMMPS_NS::PairILPGrapheneHBNOpt::calc_normal(int, int, int*, int, double*, double (*)[3], double (*)[3][3])’
>    38 |   void calc_normal(int i, int itype, int *ILP_neigh, int nneigh, double *normal, double (*dnormdri)[3],
>       |        ^~~~~~~~~~~
> In file included from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:23,
>                  from /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.cpp:22:
> /home/akohlmey/compile/lammps/src/INTERLAYER/pair_ilp_graphene_hbn.h:81:16: warning: ‘virtual void LAMMPS_NS::PairILPGrapheneHBN::calc_normal()’ was hidden [-Woverloaded-virtual=]
>    81 |   virtual void calc_normal();
>       |                ^~~~~~~~~~~
> /home/akohlmey/compile/lammps/src/OPT/pair_ilp_graphene_hbn_opt.h:38:8: note:   by ‘LAMMPS_NS::PairILPGrapheneHBNOpt::calc_normal(int, int, int*, int, double*, double (*)[3], double (*)[3][3])’
>    38 |   void calc_normal(int i, int itype, int *ILP_neigh, int nneigh, double *normal, double (*dnormdri)[3],
>       |        ^~~~~~~~~~~
> ```

@akohlmey Thank you for bringing up this issue. I have discussed it with Dr. Duan and we come up with a simple solution: renaming the "calc_normal" function in pair _ilp_graphene_opt.cpp/h to "calc_atom_normal". What are your thoughts on this proposed solution?

akohlmey commented:
> What are your thoughts on this proposed solution?

That is fine with me. It will definitely avoid the ambiguity and thus silence the warning.

oywg11 commented:
> > @run-towards-the-future @dxhisboy will update the optimized version soon.
> 
> Are these changes included? Is there anything still planned or pending from your side? Or is the pull request now ready for a final review? Please let us know.

@akohlmey Yes, we confirm that all the codes are updated and the pull request is ready for a final review now. Thanks!

oywg11 commented:
> > What are your thoughts on this proposed solution?
> 
> That is fine with me. It will definitely avoid the ambiguity and thus silence the warning.

Sure, Thanks!

akohlmey commented on doc/src/pair_aip_water_2dm.rst:
```diff
@@ -0,0 +1,165 @@
+a continuous cutoff (up to third derivative) for interatomic separations
+larger than :math:`r_c` :doc:`pair_style ilp_graphene_hbn <pair_ilp_graphene_hbn>`.
+
+It is important to include all the pairs to build the neighbor list for
```
"include all the pairs" of what?
What neighbor list does this refer to?
Does this mean that a user may need to use a particular "special_bonds" setting (e.g. special_bonds lj/cut 1.0e-100 1.0e-100 1.0e-100) for pair styles that do not require Kspace?

oywg11 replied:
To calculate the atomic normal vectors, a full neighbor list is required. In the case of a hydrogen atom, its normal vector can be determined by identifying its neighboring oxygen atom and applying the following definition: the normal vector of a hydrogen atom is assumed to lie along the corresponding oxygen-hydrogen bond.

Since it belongs to the detail of the codes, this sentence can be deleted here.

akohlmey commented on doc/src/pair_aip_water_2dm.rst:
```diff
@@ -0,0 +1,165 @@
+.. note::
+
+   Since each water molecule contains one oxygen atom and two hydrogen atoms,
+   a new definition is proposed (see In :ref:`(Feng) <Feng>`),the atomic
```
"a new definition" of what? in comparison to what?

oywg11 replied:
A new definition is proposed for determining the atomic normal vectors of water molecules, which differs from the commonly used methods for calculating normal vectors in 2D layered materials.

akohlmey commented on doc/src/pair_aip_water_2dm.rst:
```diff
@@ -0,0 +1,165 @@
+The parameter file (e.g. COH.aip.water.2dm), is intended for use with *metal*
+:doc:`units <units>`, with energies in meV. Two additional parameters,
+*S*, and *rcut* are included in the parameter file. *S* is designed to
+facilitate scaling of energies. *rcut* is designed to build the neighbor
```
Here is again a reference to "the neighbor" list, but this sounds different from the neighbor lists (there can be multiple!) that LAMMPS computes. So for clarity, this needs to be explained.

oywg11 replied:
The procedure for constructing the neighbor list for calculating atomic normal vectors is similar to the neighbor lists used in the LAMMPS computer code, as explained earlier. However, it should be noted that this is a sub-neighbor list specifically designed for the calculation of normal vectors. The parameter "Rcut" determines the cutoff distance used to determine the number of neighbors included in the calculation of the normal vectors.

akohlmey commented on doc/src/pair_aip_water_2dm.rst:
```diff
@@ -0,0 +1,165 @@
+   are fitted with taper function by setting the cutoff equal to 16.0
+   Angstrom.  Using different cutoff or taper function should be careful.
+   These parameters provide a good description in both short- and long-range
+   interaction regimes. This feature is essential for simulations in high pressure
```
Which feature?

oywg11 replied:
The feature that the force field can provide a good description in both short- and long-range interaction regimes.

akohlmey committed:
various documentation updates

- spelling fixes
- formatting conventions
- grammar fixes and clarifications
- full integration into manual build procedure

akohlmey committed:
update .gitignore

akohlmey committed:
correctly indicate OPT package version of pair style

akohlmey commented:
@oywg11 I made some updates to the documentation. Mostly formatting and formal things like adding a reference to the pair_style.rst and Commands_pair.rst files. But there are also some updates to spelling and grammar, where the change was obvious to me. There are, however, a few points that were not clear to me, so please see my questions/comments. You can either try to answer on GitHub and I will update the docs accordingly or edit the .rst file directly. Also, if any of my changes are incorrect, please update as needed or post a note here.

oywg11 commented:
> @oywg11 I made some updates to the documentation. Mostly formatting and formal things like adding a reference to the pair_style.rst and Commands_pair.rst files. But there are also some updates to spelling and grammar, where the change was obvious to me. There are, however, a few points that were not clear to me, so please see my questions/comments. You can either try to answer on GitHub and I will update the docs accordingly or edit the .rst file directly. Also, if any of my changes are incorrect, please update as needed or post a note here.

@akohlmey Thanks for the careful check. I have answered all your questions/comments. Please feel free to let me know if anything is still unclear or if you need further clarification on any aspect.

akohlmey committed:
remove unused include and related statements

akohlmey committed:
update/clarify docs

akohlmey committed:
fix versionadded macro so it is detected when actual version is added

sjplimp commented:
@akohlmey Is the (new?) dependence of the OPT and INTERLAYER packages on each other correctly setup ?

akohlmey commented:
> @akohlmey Is the (new?) dependence of the OPT and INTERLAYER packages on each other correctly setup ?

Yes, since commit da520bf8b7998ddbaafbe67e3a6eb57aab15f301 from 2022-04-22. This is not new. It was added with PR #3228.
